pylint/checkers/base.py

```diff
@@ -1488,14 +1483,14 @@ def _check_not_in_finally(self, node, node_name, breaker_classes=()):
         if not self._tryfinallys:
             return
         # the node could be a grand-grand...-children of the try...finally
-        _parent = node.parent
         _node = node
-        while _parent and not isinstance(_parent, breaker_classes):
-            if hasattr(_parent, "finalbody") and _node in _parent.finalbody:
+        for parent in node.node_ancestors():
+            if isinstance(parent, breaker_classes):
+                break
+            if hasattr(parent, "finalbody") and _node in parent.finalbody:
                 self.add_message("lost-exception", node=node, args=node_name)
                 return
-            _node = _parent
-            _parent = _node.parent
+            _node = parent
```

Comment:
Iterating through `node.parent` and `node.parent.parent` seems to be more common than I thought. Maybe we should add a utils method for it, this time in `pylint/checkers/utils.py`. Something like `walk_node_parents` or `recurse_node_parents`?

Reply:
There's another common variation where the last ancestor satisfying some condition is returned. I'm not sure if that needs another method or if there's a different pattern for that kind of thing.

Reply:
I think a generator ``n for n in node.node_ancestor if satisfy_condition(n)`` is the best solution in term of readability for those.

Reply:
Yet another variation is looping over a node's siblings rather than its ancestors.

Reply:
If we think about this even longer, I'm sure we would be able to come up with further variations.

Personally, I would like to finish this PR soon. It has taken a bit too much time already.
To keep the spirit of @Pierre-Sassoulas last comment https://github.com/PyCQA/pylint/pull/5005#discussion_r710195942, I suggest we keep these three instance as they are now. It works, so why change it. I understand that probably isn't what you were hoping for @nickdrozd, but I believe this will be the easiest (and safest) solution.

--
I'm referring only to those that use child + parent recursively.